flink-cdc-runtime/src/main/java/com/ververica/cdc/runtime/operators/schema/encoding/SchemaEncodingFunction.java

```diff
@@ -0,0 +1,273 @@
+/*
+ * Copyright 2023 Ververica Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.ververica.cdc.runtime.operators.schema.encoding;
+
+import org.apache.flink.api.common.functions.RichMapFunction;
+import org.apache.flink.api.common.state.MapState;
+import org.apache.flink.api.common.state.MapStateDescriptor;
+import org.apache.flink.configuration.Configuration;
+
+import com.ververica.cdc.common.event.AddColumnEvent;
+import com.ververica.cdc.common.event.AlterColumnTypeEvent;
+import com.ververica.cdc.common.event.CreateTableEvent;
+import com.ververica.cdc.common.event.DataChangeEvent;
+import com.ververica.cdc.common.event.DropColumnEvent;
+import com.ververica.cdc.common.event.Event;
+import com.ververica.cdc.common.event.RenameColumnEvent;
+import com.ververica.cdc.common.event.SchemaChangeEvent;
+import com.ververica.cdc.common.event.TableId;
+import com.ververica.cdc.common.schema.Column;
+import com.ververica.cdc.common.schema.PhysicalColumn;
+import com.ververica.cdc.common.schema.Schema;
+import com.ververica.cdc.common.types.DataType;
+import com.ververica.cdc.runtime.event.SchemaAwareDataChangeEvent;
+
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+import static com.ververica.cdc.common.utils.Preconditions.checkNotNull;
+
+/**
+ * Function for encoding {@link Schema} into {@link DataChangeEvent} and generate {@link
+ * SchemaAwareDataChangeEvent}.
+ */
+public class SchemaEncodingFunction extends RichMapFunction<Event, Event> {
+    private final MapStateDescriptor<TableId, Schema> schemaStateDescriptor =
+            new MapStateDescriptor<>("schemas", TableId.class, Schema.class);
+    private SchemaHolder schemaHolder;
+
+    @Override
+    public void open(Configuration parameters) throws Exception {
+        schemaHolder = new SchemaHolder(getRuntimeContext().getMapState(schemaStateDescriptor));
+        super.open(parameters);
+    }
+
+    @Override
+    public Event map(Event event) throws Exception {
+        if (event instanceof SchemaChangeEvent) {
+            schemaHolder.applySchemaChange(((SchemaChangeEvent) event));
+            return event;
+        } else if (event instanceof DataChangeEvent) {
+            DataChangeEvent dataChangeEvent = (DataChangeEvent) event;
+            return new SchemaAwareDataChangeEvent(
```

Comment:
Should we add a quick check  like schema column size and dataChangeEvent size should be equals here ?